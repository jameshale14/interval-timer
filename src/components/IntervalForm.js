import React from 'react'
import PropTypes from 'prop-types'

export default class IntervalForm extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      stepError: undefined,
      saveError: undefined,
      name: props.interval ? props.interval.name : '',
      steps: props.interval ? props.interval.steps : [],
      newStep: {
        type: 'Activity',
        name: '',
        duration: ''
      }
    }
  }

  handleAddStep = (e) => {
    e.preventDefault()
    if (this.state.newStep.type == '' || this.state.newStep.name == '' || this.state.newStep.duration == 0) {
      this.setState(() => ({ stepError: 'Please fill in all fields' }))
    } else {
      this.setState((prevState) => {
        const newStep = { ...this.state.newStep }
        return {
          steps: [...prevState.steps, newStep],
          stepError: undefined
        }
      })
      // this.setState(() => ({  }))
    }
  }

  onTypeChange = (e) => {
    const type = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.type = type
      return { newStep }
    })
  }

  onNameChange = (e) => {
    const name = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.name = name
      return { newStep }
    })
  }

  onIntervalNameChange = (e) => {
    const name = e.target.value
    this.setState(() => ({ name }))
  }

  onDurationChange = (e) => {
    const duration = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.duration = duration
      return { newStep }
    })
  }

  handleOnSave = (e) => {
    e.preventDefault()

    if (!this.state.name || this.state.steps.length === 0) {
      this.setState(() => ({ saveError: 'Please provide a routine name and interval steps.' }))
    } else {
      this.setState(() => ({ saveError: undefined }))

      const interval = {
        name: this.state.name,
        steps: this.state.steps
      }

      this.props.onSubmit(interval)
    }

  }

  handleRemoveStep = (e) => {
    e.preventDefault()
    const stepIndex = parseInt(e.currentTarget.value)
    this.setState((prevState) => {
      const steps = prevState.steps.filter((step, index) => {
        return index != stepIndex
      })

      return ({
        steps
      })

    })
  }

  render() {
    return (
      <div>
        <button id='saveInterval' onClick={this.handleOnSave}>Save</button>
        {this.state.saveError && <p>{this.state.saveError}</p>}
        <input
          type='text'
          placeholder='name'
          value={this.state.name}
          onChange={this.onIntervalNameChange}
        />
        {
          this.state.steps.map((step, stepIndex) => {
            return (
              <div key={stepIndex} >
                <p>Type: {step.type}</p>
                <p>Name: {step.name}</p>
                <p>Duration: {step.duration}</p>
                <button onClick={this.handleRemoveStep} value={stepIndex}>Remove</button>
              </div>
            )
          })
        }
        <div>
          {this.state.stepError && <p>{this.state.stepError}</p>}
          <form onSubmit={this.handleAddStep}>
            <select
              value={this.state.newStep.type}
              onChange={this.onTypeChange}
              onBlur={this.onTypeChange}
            >
              <option value='Activity'>Activity</option>
              <option value='Rest'>Rest</option>
            </select>

            <input
              type='text'
              placeholder='Step name'
              value={this.state.newStep.name}
              onChange={this.onNameChange}
            />
            <input
              type='number'
              placeholder='Duration (seconds)'
              step='1'
              value={this.state.newStep.duration}
              onChange={this.onDurationChange}
            />
            <button onClick={this.handleAddStep}>Add Step</button>
          </form>
        </div>
      </div>
    )
  }
}

IntervalForm.propTypes = {
  interval: PropTypes.object,
  onSubmit: PropTypes.func
}
