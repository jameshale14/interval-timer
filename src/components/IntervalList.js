import React from 'react'
import { connect } from 'react-redux'

export const IntervalList = (props) => (
  <div>
    {props.intervals.map((interval) => {
      return (<div>{interval.name}</div>)
    })}
  </div>
)

const mapStateToProps = (state) => ({
  intervals: []
})

export default connect(mapStateToProps)(IntervalList)