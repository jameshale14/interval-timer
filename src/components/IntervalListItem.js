import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

export const IntervalListItem = ({ id, name }) => (
  <Link to={`/edit/${id}`}>
    <h3>{name}</h3>
  </Link>
)

IntervalListItem.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string
}

export default IntervalListItem