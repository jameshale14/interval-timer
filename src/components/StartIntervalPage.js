/*eslint-env browser*/

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

export class StartIntervalPage extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      currentStepIndex: 0,
      currentStepName: undefined,
      currentTimeRemaining: undefined,
      isRunning: false
    }
    this.interval = undefined
  }

  componentDidMount() {
    document.title = 'Start Interval | Interval Timer'
  }

  componentDidUpdate() {
    document.title = 'Start Interval | Interval Timer'
  }

  initialise = (stepIndex) => {
    // const currentStepIndex = this.state.currentStepIndex
    this.setState(() => ({
      currentStepName: this.props.interval.steps[stepIndex].name,
      currentTimeRemaining: this.props.interval.steps[stepIndex].duration
    }))
  }

  runTimer = () => {
    if (!this.state.isRunning) {
      return
    }

    const currTime = this.state.currentTimeRemaining - 1
    this.playSound(currTime)

    if (this.state.currentTimeRemaining == 1) {
      if (this.state.currentStepIndex < (this.props.interval.steps.length - 1)) {
        this.setState((prevState) => ({
          currentStepIndex: (prevState.currentStepIndex + 1)
        }))
        this.initialise(this.state.currentStepIndex)
      } else {
        window.clearInterval(this.interval)
        this.setState(() => ({ isRunning: false }))
      }

    } else {
      // const currTime = this.state.currentTimeRemaining

      this.setState((prevState) => ({
        currentTimeRemaining: (prevState.currentTimeRemaining - 1)
      }))

      // this.playSound(currTime)

    }
  }

  playSound = (timeRemaining) => {
    if (timeRemaining > 3) {
      return
    }

    if (timeRemaining == 0) {
      const AudioContext = window.AudioContext || window.webkitAudioContext
      const context = new AudioContext()
      const g = context.createGain()
      const o = context.createOscillator()
      o.type = 'sine'
      o.connect(g)
      o.frequency.value = 659.26
      g.connect(context.destination)
      o.start(0)
      g.gain.exponentialRampToValueAtTime(0.00001, context.currentTime + 2.5)

    } else {
      const AudioContext = window.AudioContext || window.webkitAudioContext
      const context = new AudioContext()
      const g = context.createGain()
      const o = context.createOscillator()
      o.type = 'sine'
      o.connect(g)
      o.frequency.value = 440.0
      g.connect(context.destination)
      o.start(0)
      g.gain.exponentialRampToValueAtTime(0.00001, context.currentTime + 1)

    }
  }

  handleStart = (e) => {
    e.preventDefault()
    if (this.state.isRunning) {
      return
    }
    if (this.state.currentTimeRemaining == undefined) {
      this.initialise(0)
    }
    this.setState(({ isRunning: true }))
    this.interval = setInterval(this.runTimer, 1000)

  }

  handleStop = (e) => {
    e.preventDefault()

    window.clearInterval(this.interval)
    this.setState(() => ({ isRunning: false }))
  }

  handleRestartStep = (e) => {
    e.preventDefault()

    this.initialise(this.state.currentStepIndex)
    this.setState(({ isRunning: true }))
    this.interval = setInterval(this.runTimer, 1000)

  }

  handleRestartWholeInterval = (e) => {
    e.preventDefault()

    this.setState(() => ({ currentStepIndex: 0 }))
    this.initialise(0)
    this.setState(({ isRunning: true }))
    this.interval = setInterval(this.runTimer, 1000)

  }

  render() {
    return (
      <div>
        <h3 data-testid='interval-name'>{this.props.interval.name}</h3>
        <h1 data-testid='step-name'>{this.state.currentStepName}</h1>
        <h1 data-testid='time-remaining'>{this.state.currentTimeRemaining}</h1>
        <button onClick={this.handleStart}>Start</button>
        <button onClick={this.handleStop}>Stop</button>
        {!this.state.isRunning && this.state.currentStepName != undefined &&
          <>
            <button onClick={this.handleRestartStep}>Restart Step</button>
            <button onClick={this.handleRestartWholeInterval}>Restart Interval</button>
          </>
        }
      </div>
    )
  }

}

StartIntervalPage.propTypes = {
  interval: PropTypes.object
}

const mapStateToProps = (state, props) => ({
  interval: state.intervals.find((interval) => interval.id === props.match.params.id)
})

export default connect(mapStateToProps)(StartIntervalPage)