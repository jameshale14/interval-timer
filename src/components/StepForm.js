import React from 'react'
import PropTypes from 'prop-types'

export default class StepForm extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      error: undefined,
      isRestType: false,
      type: 'Activity',
      name: '',
      duration: 0
    }
  }

  onTypeChange = (e) => {
    const type = e.target.value
    const isRestType = type == 'Rest' ? true : false
    const name = isRestType ? 'Rest' : ''
    this.setState(() => ({ type, isRestType, name }))
  }

  onNameChange = (e) => {
    const name = e.target.value
    this.setState(() => ({ name }))
  }

  onDurationChange = (e) => {
    const duration = e.target.value
    this.setState(() => ({ duration }))
  }

  handleAddStep = (e) => {
    e.preventDefault()
    if (this.state.type == '' || this.state.name == '' || this.state.duration == 0) {
      this.setState(() => ({ error: 'Please fill in all fields' }))
    } else {
      this.setState(() => ({ error: undefined }))
      const newStep = {
        name: this.state.name,
        type: this.state.type,
        duration: this.state.duration
      }
      this.props.onSubmit(newStep)
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleAddStep}>
          <label htmlFor='stepType' >Step Type</label>
          <select
            id='stepType'
            value={this.state.type}
            onChange={this.onTypeChange}
            onBlur={this.onTypeChange}
          >
            <option value='Activity'>Activity</option>
            <option value='Rest'>Rest</option>
          </select>

          <label htmlFor='stepName'>Step Name</label>
          <input
            id='stepName'
            type='text'
            placeholder='Step name'
            value={this.state.name}
            onChange={this.onNameChange}
            disabled={this.state.isRestType}
          />

          <label htmlFor='stepDuration'>Step Duration</label>
          <input
            id='stepDuration'
            type='number'
            placeholder='Duration (seconds)'
            step='1'
            value={this.state.duration}
            onChange={this.onDurationChange}
          />
          <button onClick={this.handleAddStep}>Add Step</button>
        </form>
      </div>
    )
  }
}

StepForm.propTypes = {
  onSubmit: PropTypes.func
}