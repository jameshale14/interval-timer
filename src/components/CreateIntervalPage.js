import React from 'react'
import { connect } from 'react-redux'
import IntervalForm from './IntervalForm'

export class CreateIntervalPage extends React.Component {
  onSubmit = (interval) => {
    this.props.startCreateInterval(interval)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <IntervalForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startCreateInterval: (interval) => dispatch(startCreateInterval(interval))
})

export default connect(undefined, mapDispatchToProps)(CreateIntervalPage)