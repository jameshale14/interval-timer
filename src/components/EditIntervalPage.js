import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import IntervalForm from './IntervalForm'
import { startUpdateInterval } from '../actions/intervals'

export class EditIntervalPage extends React.Component {

  onSubmit = (interval) => {
    this.props.startUpdateInterval(this.props.interval.id, interval)
    this.props.history.push('/')
  }

  render() {
    return (
      <>
        <h3>Edit Interval</h3>
        <div>
          <IntervalForm onSubmit={this.onSubmit} interval={this.props.interval} />
        </div>
      </>
    )
  }
}

EditIntervalPage.propTypes = {
  interval: PropTypes.object,
  startUpdateInterval: PropTypes.func,
  history: PropTypes.object
}

const mapStateToProps = (state, props) => ({
  interval: state.intervals.find((interval) => interval.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
  startUpdateInterval: (id, interval) => dispatch(startUpdateInterval(id, interval))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditIntervalPage)