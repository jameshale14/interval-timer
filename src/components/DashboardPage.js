import React from 'react'
import { Link } from 'react-router-dom'
import IntervalList from './IntervalList'

const DashBoardPage = () => (
  <div>
    <Link className='button' to='/create'>New Interval</Link>
    <IntervalList />
  </div>
)

export default DashBoardPage 