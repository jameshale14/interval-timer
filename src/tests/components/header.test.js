import React from 'react'
import { fireEvent, cleanup } from 'react-testing-library'
import { renderWithRouter } from '../utils/render'
import { Header } from '../../components/Header'

let startLogout

beforeEach(() => {
  startLogout = jest.fn()
})

afterEach(cleanup)

test('should render Header correctly', () => {
  const { container } = renderWithRouter(<Header startLogout={startLogout} />)
  expect(container).toMatchSnapshot()
})

test('should call startLogout on button click', () => {
  const { getByText } = renderWithRouter(<Header startLogout={startLogout} />)
  const logoutButton = getByText('Logout')
  fireEvent.click(logoutButton)
  expect(startLogout).toHaveBeenCalled()
})