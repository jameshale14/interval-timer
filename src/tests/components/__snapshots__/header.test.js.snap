// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="content-container"
    >
      <div
        class="header__content"
      >
        <a
          class="header__title"
          href="/dashboard"
        >
          <h1>
            Interval Timer
          </h1>
        </a>
        <button
          class="button button--link"
        >
          Logout
        </button>
      </div>
    </div>
  </header>
</div>
`;
